Make ICE server URLs configurable

heartlabs opened the pull request:
Implements #27 

This is my first public Rust contribution so I am happy about feedback especially about API conventions and borrow rules

I guess we should also add some tests - but right now I am not sure if this is even possible without major refactoring...

johanhelsing reviewed (changes requested):
> This is my first public Rust contribution so I am happy about feedback especially about API conventions and borrow rules

Apart from some naming conventions, it looks fine :)

> I guess we should also add some tests - but right now I am not sure if this is even possible without major refactoring...

Yeah, testing would be nice, but I agree it doesn't make sense to add it in this PR. If you have any ideas about what/how to make the code more testable, we can open another issue for it :)

Thanks for the PR :)

heartlabs commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -36,6 +36,23 @@ use uuid::Uuid;
 type Packet = Box<[u8]>;
 
+#[derive(Debug)]
+pub struct MatchboxConfig {
```
Right that's better - I changed it

heartlabs commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -54,6 +71,11 @@ pub(crate) type MessageLoopFuture = Pin<Box<dyn Future<Output = ()>>>;
+    }
+
+    #[must_use]
+    pub fn custom<T: Into<String>>(room_url: T, config: MatchboxConfig) -> (Self, MessageLoopFuture) {
```
I agree with you here - changed the name

heartlabs commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -36,6 +36,23 @@ use uuid::Uuid;
 
+#[derive(Debug)]
+pub struct MatchboxConfig {
+    pub ice_server_urls: Vec<String>
```
I see - I also misunderstood that. I added some comments and also adapted the config struct(s) to better reflect that.

heartlabs commented:
I think I fixed everything you suggested. I also moved the `room_url` into the config struct because I think that looks neater. Please let me know if this also makes sense.

heartlabs commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -36,6 +36,23 @@ use uuid::Uuid;
+    fn default() -> Self {
+        MatchboxConfig {
+            ice_server_urls: vec![
+                "stun:stun.l.google.com:19302".to_string(),
```
Yes it did actually!

johanhelsing commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -36,6 +36,37 @@ use uuid::Uuid;
+/// Configuration options for an ICE server connection.
+/// See also: https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer#example
+pub struct RtcIceServerConfig {
+    /// An ICE server instance can have several URLs. Use an empty Vec to disable ICE
```
> Use an empty Vec to disable ICE

Are you sure this is allowed? MDN says it is required and should contain one or more urls.

If we want the ability to run without an ice server, I think it is the `RtcConfig.iceServers`/`ice_server_config_list` that should be empty, not the url list for the single server.

If we want a way to disable ICE, I think a better API would be to make `WebRtcSocketConfig.ice_server` an `Option<RtcIceServerConfig>` or ideally a `Vec<RtcIceServerConfig>` so we can support multiple servers as well (and it would mirror the underlying API).

Good idea to make it a struct, though. Makes it easier when/if we add TURN support.

heartlabs replied:
I see - to be honest I just tried it out and documented what I saw happening. Perhaps it only worked because I had a local-only setup. Probably it's best if I just remove that comment for now to avoid confusion

heartlabs replied:
I also like your suggestion and had similar ideas - I just wanted to keep this PR as minimal as possible and just add configurability without changing anything in the actual implementation because I don't know too much about the topic and dont't want to accidentally break something :)

johanhelsing reviewed (changes requested):
I like it :) good idea to move the `room_url` parameter!

heartlabs commented:
From my side the PR would be ready to be merged - is there anything else you think we should address in this scope?

johanhelsing commented on matchbox_socket/src/webrtc_socket/mod.rs:
```diff
@@ -36,6 +36,37 @@ use uuid::Uuid;
+/// Configuration options for an ICE server connection.
+/// See also: https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer#example
+pub struct RtcIceServerConfig {
+    /// An ICE server instance can have several URLs. Use an empty Vec to disable ICE
```
```suggestion
    /// An ICE server instance can have several URLs.
```
Yep, probably best to just remove it, and will deal with disabling/multiple in a separate pr.

heartlabs replied:
Interesting that GitHub shows it like this but I did that exact change already before your change request: https://github.com/johanhelsing/matchbox/pull/28/commits/f623ce7bb4c85261f3c5203572f2b5fc534de8be

johanhelsing commented:
> From my side the PR would be ready to be merged - is there anything else you think we should address in this scope?

Just the doc comment. Otherwise it looks great :) Thanks for working on this!

johanhelsing reviewed (approved):
Ah weird I didn't see it... Perhaps I clicked the wrong diff button.

heartlabs commented:
> > From my side the PR would be ready to be merged - is there anything else you think we should address in this scope?
> 
> Just the doc comment. Otherwise it looks great :) Thanks for working on this!

It was my pleasure :) Should I do something about the failing lints? And if yes - can I run some commands locally to reproduce all these issues?

johanhelsing commented:
Should be fixed by running `cargo fmt`.

In most editors, you can set it up to run automatically on save.

The easiest thing is probably to just run the failing cargo commands yourself, but if you have docker set up you could also install https://github.com/nektos/act if you want to run the ci setup locally.
